### PR TITLE
feat: static SFT through uv run rl

### PR DESCRIPTION
**Pull request by tim0120**
## What

Adds pure static SFT as a supported `uv run rl` configuration. The dataset is an orchestrator sampling source, not an environment, and the existing RL trainer consumes the resulting `TrainingBatch` with its CE loss.

```toml
[orchestrator.algo]
type = "sft"

[orchestrator.algo.sampling.source]
type = "dataset"
name = "org/my-sft-dataset" # messages or prompt + completion
```

Pure static runs need trainer GPUs only; they do not start policy inference. Configuring an `[inference]` block or NCCL weight broadcast for a dataset-backed run is rejected.

## How

- Renames the runtime to one `SFTAlgorithm`. Frozen-model hard distillation and static datasets are the two supported `sft` sources.
- Adds an explicitly typed dataset variant under `orchestrator.algo.sampling.source`, reusing the existing SFT dataset selection and role-mask controls.
- Adds a dataset batch producer inside the orchestrator that loads and renders rows into the existing `TrainingSample` / `TrainingBatch` contract and stamps those tokens for CE.
- Shares message/tool normalization with the standalone SFT data path instead of duplicating its row parsing.
- Selects the static path at the top of `run_orchestrator`; the normal environment/rollout path, transport, packer, and trainer are unchanged.
- Separates launching the orchestrator from launching inference in multi-node runs. Zero-inference static runs start the orchestrator on trainer rank 0, while online zero-inference runs remain invalid.
- Skips policy inference for pure static SFT. The static loop waits for the trainer's existing filesystem `STABLE` marker between updates, without changing `WeightWatcher`; shared NCCL configuration is rejected because no inference ranks can join it.
- Preserves the dataset position in orchestrator progress for deterministic checkpoint resume.
- Caps accepted rows at the smaller of the dataset's optional `max_length` and the trainer's `seq_len`, and masks position zero to match the RL trainer's shifted-token contract.
- Supports token-based static batches without requiring the rollout-only `max_inflight_rollouts` setting.

This intentionally does not add static/RL mixing, `loss_weight`, prefetching, or a second normalization policy. The trainer's existing CE normalization is unchanged. Trainer-step acknowledgement is separated from weight publication in follow-up draft #3024.

## Validation

- Focused config coverage for explicit dataset typing, rejection of unused inference and shared NCCL, and the static token-batching carve-out.
- Static multi-node config validation passed with zero inference nodes; the corresponding online config remains rejected.
- The rendered zero-inference multi-node SLURM script passes `bash -n`, launches the orchestrator on trainer rank 0, and does not reference inference URLs.
- Ruff, Python 3.12 compile, and diff checks pass on the changed paths. Full validation runs in Linux CI; the repository lock does not support this macOS host.

No new test files or README changes; documentation edits are limited to the affected SFT descriptions.

---

> [!NOTE]
> **Medium Risk**
> Introduces a parallel orchestrator training loop and deployment rules (no inference, filesystem sync only); misconfiguration is mostly caught at validate time, but step handshake and batch rendering behavior are new production paths beside rollout RL.
> 
> **Overview**
> **Static SFT through `uv run rl`** — `sft` can use `orchestrator.algo.sampling.source` with `type = "dataset"` and a Hugging Face dataset name (plus existing SFT dataset knobs like `loss_mask`), instead of only a frozen teacher endpoint. The runtime class is unified as **`SFTAlgorithm`** (replacing `SFTDistillAlgorithm`).
> 
> **Orchestrator path** — When the top-level algo is dataset-backed SFT, `run_orchestrator` branches to a new **`DatasetBatchProducer`** that loads rows, renders them with the orchestrator renderer, stamps **`TrainingSample`**s for CE, sends **`TrainingBatch`**es on the existing transport, and waits on the trainer’s filesystem **`STABLE`** marker between steps. Dataset cursor lives in orchestrator progress for resume. This path forbids **`train.env`**, eval, and mixing dataset sources into per-env algos.
> 
> **Launch / deployment** — Static runs set **`needs_inference = false`**: no policy vLLM, **`[inference]`** rejected at the RL config layer, and **filesystem-only** weight broadcast (NCCL rejected). Multi-node SLURM decouples **`launch_orchestrator`** from inference nodes so zero-inference static SFT can run the orchestrator on trainer rank 0; token batching no longer requires **`max_inflight_rollouts`** for static SFT.
> 
> **Shared plumbing** — **`resolve_sft_messages`** / **`resolve_sft_tools`** centralize SFT row parsing for both the standalone trainer dataloader and the static orchestrator path. Docs and **`configs/debug/static_sft.toml`** document the new mode.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3202bf5f2930c3566be47b3acb291ad2e78aa737. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by tim0120** [comment]
@codex review

**Comment by tim0120** [comment]
@codex review
